src/ng/compile.js

```diff
@@ -2424,6 +2426,9 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
           if (isFunction(controller.instance.$onInit)) {
             controller.instance.$onInit();
           }
+          if (isFunction(controller.instance.$onDestroy)) {
+            scope.$on('$destroy', controller.instance.$onDestroy);
```

Comment:
Two main comments here (was actually implementing this just now until I saw this):

1.  This doesn't allow late-bound `$onDestroy` methods to be added.  If it's not there at the time of controller instantiation, the lifecycle hook will not function.  Does that follow expectations, or do we need to be checking this at the time of destruction?

2.  The `this` binding in the `$onDestroy` method is going to be incorrect here due to the manner in which you're passing `$onDestroy`.  I see that you copied the unit tests from the one above it as I had done, but removed the `check` call.  You should re-introduce that portion of the test since it would have failed on this point